File: Stack/하현준/0234-palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        ListNode slow = head, fast = head;

        // find middle
        while (fast != null && fast.next != null) {
            fast = fast.next.next;
            slow = slow.next; // slow will be on a middle of the list
        }

        // reverse pointer of second half
        ListNode prev = slow;
        slow = slow.next;
        prev.next = null;
        while (slow != null) {
            ListNode temp = slow.next;
            slow.next = prev; // reverse pointer
            prev = slow; // move pointer
            slow = temp; // move pointer
        }

        // at this point, prev will point to the last node of list
        fast = head;
        slow = prev;
        while (slow != null) {
            if (fast.val != slow.val) {
                return false;
            }
            fast = fast.next;
            slow = slow.next;
        }
        return true;
    }
}

File: Leetcode/Stack/김민성/0234-palindrome-linked-list.swift
//
//  0234-palindrome-linked-list.swift
//  algo
//
//  Created by MIN SEONG KIM on 2023/03/26.
//

import Foundation


// Definition for singly-linked list.
public class ListNode {
    public var val: Int
    public var next: ListNode?
    public init() { self.val = 0; self.next = nil; }
    public init(_ val: Int) { self.val = val; self.next = nil; }
    public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
}
 
class Solution {
    func isPalindrome(_ head: ListNode?) -> Bool {
        var slow = head, fast = head, dummy: ListNode? = nil
        
        // reverse first half
        while fast != nil && fast!.next != nil {
            fast = fast!.next!.next
            
            let nextNode = slow!.next
            
            slow!.next = dummy
            
            dummy = slow
            slow = nextNode
        }
        
        // go to the starting point when length of list is odd
        if fast != nil {
            if slow == nil {
                return true
            }
            slow = slow!.next
        }
        
        // compare reversed first and second half
        while slow != nil {
            if slow!.val != dummy!.val {
                return false
            } else {
                slow = slow!.next
                dummy = dummy!.next
            }
        }
        
        return true
    }
}
